lib/thrift/parser/nimble.ex

```diff
@@ -0,0 +1,189 @@
+      |> choice()
+      |> label("punctuator")
+    end
+
+    defp literal() do
+      choice([
+        literal_with(?"),
+        literal_with(?')
+      ])
+      |> reduce({List, :to_string, []})
+      |> label("literal")
+    end
+
+    defp literal_with(char) do
+      delim = ascii_char([char])
+
+      delim
+      |> ignore()
+      |> concat(
+        choice([
+          utf8_char([?\\]) |> ignore() |> concat(delim),
+          utf8_char([]),
+          error(eos(), "expected literal delimiter ?#{[char]}"),
+          error(empty(), "expected utf8 codepoint")
+        ])
+        |> repeat_until([delim])
+      )
+      |> ignore(delim)
+    end
+
+    defp number(combinator \\ empty()) do
+      combinator
+      |> choice([
+        ascii_char([?-, ?+])
+        |> choice([
+          unsigned_number(),
+          empty()
+          |> error("expected number")
+        ])
+        |> post_traverse({__MODULE__, :__sign__, []}),
+        unsigned_number()
+      ])
+      |> label("number")
+    end
+
+    defp unsigned_number() do
+      choice([
+        hex(),
+        integer(min: 1)
+        |> choice([
+          ignore(ascii_char([?.]))
+          |> integer(min: 1)
+          |> optional(ignore(ascii_char([?E, ?e])) |> exponent()),
+          empty()
+          |> replace(0)
+          |> ignore(ascii_char([?E, ?e]))
+          |> exponent(),
+          empty()
+        ])
+      ])
+      |> optional(
+        ascii_char([?.])
+        |> ignore()
+        |> error(empty(), "expected integer fraction for significand")
+      )
+      |> post_traverse({__MODULE__, :__number__, []})
+    end
+
+    def __sign__(_rest, acc, context, _line, _offset) do
+      case acc do
+        [number, ?-] ->
+          {[-number], context}
+
+        [number, ?+] ->
+          {[number], context}
+      end
+    end
+
+    defp hex() do
+      string("0x")
+      |> ignore()
+      |> choice([
+        ascii_string([?0..?9, ?a..?f, ?A..?F], min: 1)
+        |> map({String, :to_integer, [16]}),
+        empty()
+        |> error("expected hexidecimal digit")
+      ])
+    end
+
+    defp exponent(combinator) do
+      combinator
+      |> choice([
+        choice([
+          ascii_char([?-, ?+]),
+          empty() |> replace(?+)
+        ])
+        |> integer(min: 1),
+        empty()
+        |> error("expected integer exponent")
+      ])
+    end
+
+    def __number__(_rest, acc, context, _line, _offset) do
+      case acc do
+        [_int] ->
+          {acc, context}
+
+        [fraction, int] ->
+          {[String.to_float("#{int}.#{fraction}")], context}
+
+        [exponent, exponent_sign, fraction, int] ->
+          {[String.to_float("#{int}.#{fraction}e#{[exponent_sign]}#{exponent}")], context}
+      end
+    end
+
+    def identifier() do
+      ascii_char([?a..?z, ?A..?Z, ?_])
+      |> repeat(ascii_char([?a..?z, ?A..?Z, ?_, ?0..?9]))
+      |> reduce({List, :to_atom, []})
```

Comment:
Consider extracting these three lines into their own function because we repeat them below.